### PR TITLE
Replace `serve` with `ServeHandler`

### LINKED ISSUE #384: Overhaul the serve method - write a Handler backed by the asset graph
We need to be able to do some path translation for assets that would come from locations on disk other than their apparent path.

The [comment](https://github.com/dart-lang/build/blob/dead012b3590c76d7575b24723afb490a04aa3b3/build_runner/lib/src/generate/build.dart#L129-L131) is misleading here. We talk about overriding the `requestHandler` to get a different behavior than a server on `address:port` - but this happens regardless. The "custom behavior" you can override is reading files from disk - you can't change how they are served. We have a bug where we don't properly pass through `requestHandler`, but this is fine - changing that isn't something we'd want to support anyway. If anything we should have our request handler go through 

I think we should change our signature to something like:

```dart
abstract class BuildHandler {
  Stream<BuildResult> get builds;
  Future<Reponse> handle(Request request);
}
BuildHandler buildHandler(...);
```

Usage like:

```dart
import 'package:shelf/shelf.dart';
import 'package:build_runner/build_runner.dart;

import 'phases.dart';

void main() {
  // Here `serve` is coming from package:shelf rather than package:build_runner
  serve(buildHandler(phases).handle, address, port);
}
```

### PR DESCRIPTION
**Pull request by natebosch**
Closes #384

It will now be the responsibility of the calling script to start the
server with shelf.

- Add `BuildHandler` class which implement `BuildState` and adds a shelf
  handler method.
- Add a handler backed by an `AssetReader`.
- Make `WatchImpl` public again and expose the necessary properties for
  creating the new handler.

### CODE REVIEW

**Comment by natebosch** [comment]
Getting some early eyes on this. Will add tests.

File: build_runner/lib/src/generate/build.dart [hunk 1]
```diff
@@ -121,8 +121,9 @@ Stream<BuildResult> watch(List<BuildAction> buildActions,
 /// By default a static server will be set up to serve [directory] at
```
Comment (jakemac53) [inline build_runner/lib/src/generate/build.dart]: will need to update the docs :)

File: build_runner/lib/src/server/server.dart [hunk 2]
```diff
@@ -4,45 +4,56 @@
...
+        : new AssetId(_rootPackage, p.joinAll(pathSegments));
+    try {
+      if (!await _reader.canRead(assetId)) {
+        return new Response.notFound('Not Found');
+      }
+    } on ArgumentError catch (_) {
+      return new Response.notFound('Not Found');
+    }
+    var bytes = await _reader.readAsBytes(assetId);
+    var headers = {HttpHeaders.CONTENT_LENGTH: '${bytes.length}'};
```
Comment (jakemac53) [inline build_runner/lib/src/server/server.dart]: We are going to want to add more headers here as well, to handle some form of caching (probably just etag based?)

File: build_runner/lib/src/generate/build.dart [hunk 3]
```diff
@@ -121,8 +121,9 @@ Stream<BuildResult> watch(List<BuildAction> buildActions,
...
 /// [address]:[port], but instead a [requestHandler] may be provided for custom
 /// behavior.
-Stream<BuildResult> serve(List<BuildAction> buildActions,
+Future<BuildHandler> buildHandler(List<BuildAction> buildActions,
```
Comment (jakemac53) [inline build_runner/lib/src/generate/build.dart]: I am not sure about the name here, it sounds more like its handling builds than requests :).

File: build_runner/lib/src/generate/build.dart (hunk 1 above)
Comment (natebosch) [reply build_runner/lib/src/generate/build.dart]: Done

File: build_runner/lib/src/generate/build.dart (hunk 3 above)
Comment (natebosch) [reply build_runner/lib/src/generate/build.dart]: Change to ServeHandler and dropped this method in favor of having only `watch`

File: build_runner/lib/src/server/server.dart (hunk 2 above)
Comment (natebosch) [reply build_runner/lib/src/server/server.dart]: yeah definitely but I think we can wait until we have some experience using this and add it then. We'll be heavily influence by the types of optimizations we make using the AssetGraph

**Comment by natebosch** [comment]
If we wanted to we *could* make `watch` return a synchronous `ServeHandler` and hid more of the asynchrony underneath since it only exposes asynchronous methods. Let me know if you want me to try it. See `watch_test.dart` for the type of work that needs to be done by clients otherwise if they want to keep a synchronous access to `Stream<BuildResult>`

**Pull request merged**